docs/proposal/20211207-delegated-storage.md

```diff
@@ -0,0 +1,1186 @@
+        1. S3 buckets (e.g. for external Terraform state)
+
+- Ability to manage such TypeInstances manually (via CLI and maybe UI) and as a part of Action
+    - Support such dynamic TypeInstance creation, update, and deletion
+    - Define GraphQL API and Implementation workflow syntax
+
+- Support for automatic TypeInstance creation and deletion
+    - React on events to create/delete such TypeInstances (e.g. Kubernetes events for any change for Flux's HelmReleases).
+
+- Support extensibility for upcoming backends
+
+Also, the additional, nice-to-have goals are:
+
+- Remove Neo4j dependency from Local Hub while preserving TypeInstance metadata, such as relations
+
+### Non-goal
+
+- Support external back-ends for Capact manifest storage (Public Hub)
+- Remove Neo4j dependency from Public Hub
+
+## Assumptions
+
+1. Content Developer should be able to:
+    1. Write manifests without specifying a storage backend (use default one configured by System Administrator). In this case, a static TypeInstance value is stored in the default storage backend.
+    1. Specify a specific storage backend as a part of a given Implementation. This case supports both static and dynamic TypeInstance values.
+1. System Administrator can configure default backend storage for static values.
+1. There are two different cases when it comes CRUD operations on TypeInstances:
+    1. CRUD operations on TypeInstance actually manages external resource (e.g. Vault). That is, CRUD operations on TypeInstances in Local Hub actually creates, updates and deletes a given resource.
+    1. CRUD operations on TypeInstance represents external resources managed in different way (e.g. by running Helm.install). That is, CRUD operations on TypeInstances in Local Hub actually registers, unregisters and updates references for external state without changing them.
+
+## Prerequisites
+
+1. We implement these two Type features:
+    - https://capact.io/docs/feature/type-features/#additional-references-to-parent-nodes
+    - https://capact.io/docs/feature/type-features#find-types-based-on-prefix-of-parent-nodes
+1. We add `TypeInstance.spec.backend` field (string)
+1. [Add TypeInstance `alias` metadata field](https://github.com/capactio/capact/issues/579)
+
+## Registering storage backends
+
+1. For every storage backend, we create a dedicated Type:
+
+    ```yaml
+    ocfVersion: 0.0.1
+    revision: 0.0.1
+    kind: Type
+    metadata:
+      path: cap.type.helm.storage
+    spec:
+      additionalRefs:
+        - "cap.core.type.hub.storage" 
+      jsonSchema:
+        value: # JSON schema with:
+          {
+          "$schema": "http://json-schema.org/draft-07/schema",
+          "type": "object",
+          "title": "The root schema",
+          "required": [
+            "url",
+            "additionalParametersSchema"
+          ],
+          "properties": {
+            "url": { # url of hosted app, which implements storage backend ProtocolBuffers interface.
```

Comment:
That's something I didn't describe explicitly, but I believe the URL should be always internal and that's why we don't need to extend that with auth. 
Assumption: such services are internal and service accepts connection only from Hub.

I will write it somewhere

Reply:
maybe it should also be deployed in a given namespace 🤔  it will be probably hard to achieve. I'm thinking what if this is open URL and cluster where Capact is installed is used also by others? Someone can save/get/etc what he wants. It will be good to scope it somehow with rbac then 🤔  e.g. auth + secret in a system namespace.

Reply:
As discussed, I will write about:
- [network policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/)
- refer to issue with user authz - only cluster admin can run such Actions
- to think about: backend installs inside `capact-system`